core/types/calls.cc

```diff
@@ -962,6 +970,107 @@ DispatchResult dispatchCallSymbol(const GlobalState &gs, const DispatchArgs &arg
+                auto flags = data->arguments[i].flags;
+                if (flags.isBlock || flags.isRepeated) {
+                    continue;
+                }
+                if (!flags.isDefault) {
+                    areOtherParamsDefault = false;
+                    break;
+                }
+            }
+
+            auto maybeKwargIdx =
+                ait - args.args.begin() < args.args.size() ? ait - args.args.begin() : args.args.size() - 1;
+            auto allArgsProcessed = false;
+
+            // `hasKwsplat` can be false positive, because it relies on how we parse positional and keyword args
+            auto actuallyHasKwsplatParam = absl::c_any_of(
+                data->arguments, [](const auto &arg) { return arg.flags.isKeyword && arg.flags.isRepeated; });
+
+            // Check if remaining keyword arg names are matching keyword param names
+            // If kwargs splat exists, all kwargs are allowed and type mismatches will be reported later in the method
+            if (!actuallyHasKwsplatParam) {
+                UnorderedSet<NameRef> keywordArgsNames;
+                for (const auto &param : data->arguments) {
+                    if (!param.flags.isKeyword) {
+                        continue;
+                    }
+                    keywordArgsNames.insert(param.name);
+                }
+
+                if (keywordArgsNames.size() != 0) {
+                    {
+                        auto kwit = args.args.begin() + args.numPosArgs;
+                        auto kwend = args.args.begin() + args.numPosArgs + numKwargs;
+
+                        allArgsProcessed = true;
+                        while (kwit != kwend) {
+                            auto &key = *kwit++;
+                            if (isa_type<NamedLiteralType>(key->type)) {
+                                auto keyLiteral = cast_type_nonnull<NamedLiteralType>(key->type);
+
+                                if (keyLiteral.literalKind == NamedLiteralType::LiteralTypeKind::Symbol &&
+                                    keywordArgsNames.find(keyLiteral.asName()) != keywordArgsNames.end()) {
+                                    kwit++;
+
+                                } else {
+                                    allArgsProcessed = false;
+                                    break;
+                                }
+                            }
+                        }
+                    }
+                }
+            }
+
+            if (!allArgsProcessed &&
+                (
+                    // there are keyword args remaining and we expect the current argument to be a hash,
+                    // but it's not a hash => report implicit conversion
+                    (isTyped(probablyFirstKwarg.type) && isHash(gs, probablyFirstKwarg.type) &&
+                     isTyped(args.args[maybeKwargIdx]->type) && !isHash(gs, args.args[maybeKwargIdx]->type) &&
```

Comment:
It seems odd to me that we might ask `isHash` on a value that is just the inlined `Symbol` argument corresponding to a keyword arg's name.

Reply:
We can replace it with a check for is it a `Symbol`